src/services/download-service.ts

```diff
@@ -12,123 +13,153 @@ class DownloadService {
+    metadata!.imageDownloadUrl = await this.handleImageDownload();
+    const downloadUrls: Map<string, string> = await this.getDownloadUrls(metadata);
+
+    // assign the download urls to the assets by id
+    metadata?.captions.forEach(captionAsset => (captionAsset.downloadUrl = downloadUrls.get(captionAsset.id)!));
+    metadata?.flavors.forEach(flavorAsset => (flavorAsset.downloadUrl = downloadUrls.get(flavorAsset.id)!));
+    metadata?.attachments.forEach(attachmentAsset => (attachmentAsset.downloadUrl = downloadUrls.get(attachmentAsset.id)!));
+
+    metadata!.fileName = this.getFilename(metadata);
+    return metadata;
+  }
 
-    const partnerPart = `/p/${partnerId}`;
-    const entryIdPart = `/entryId/${id}`;
-    const flavorParamIdPart = flavorParamId !== null ? `/flavorParamId/${flavorParamId}` : '';
-    const flavorIdPart = flavorId !== null ? `/flavorId/${flavorId}` : '';
-    const ksPart = ks ? `/ks/${ks}` : '';
-    const protocolPart = `/protocol/${location.protocol.split(':')[0]}`;
+  private async getDownloadUrls(metadata: DownloadMetadata): Promise<Map<string, string>> {
+    const ks = this.player.config.session?.ks || '';
+    let urls = new Map<string, string>();
+
+    const data = await this.player.provider.doRequest(
+      [{loader: DownloadUrlLoader, params: {flavors: metadata?.flavors, captions: metadata?.captions, attachments: metadata?.attachments}}],
+      ks
+    );
+    if (data && data.has(DownloadUrlLoader.id)) {
+      const urlsLoader = data.get(DownloadUrlLoader.id);
+      urls = urlsLoader?.response?.urls;
+    }
 
-    return `${cdnUrl}${partnerPart}/playManifest${entryIdPart}${protocolPart}${ksPart}${flavorParamIdPart}${flavorIdPart}/format/download`;
+    return urls;
   }
 
-  async getDownloadMetadata(config: DownloadConfig): Promise<DownloadMetadata> {
-    if (!(this.isPlatformSupported() && this.isEntrySupported())) {
-      return null;
+  private async handleImageDownload(): Promise<string> {
+    if (!this.player.isImage()) return '';
+    let imageRequestUrl = this.player.sources.downloadUrl;
+    if (!imageRequestUrl) {
+      return await this.getImageDownloadMetadata();
     }
-
-    const requestUrl = this.getRequestUrl(config);
-    if (!requestUrl) {
-      return null;
+    if (!imageRequestUrl.includes('/ks/')) {
+      const ks = this.player.config.session?.ks;
+      imageRequestUrl = ks ? `${imageRequestUrl}/ks/${ks}` : imageRequestUrl;
     }
-
     try {
-      const response = await fetch(requestUrl, {method: 'HEAD'});
-      if (!response.ok) {
-        return null;
-      }
+      const response = await fetch(imageRequestUrl, {method: 'HEAD'});
+      return response.url;
+    } catch (e: any) {
+      this.logger.warn('Failed to get image file from url: ', imageRequestUrl);
+      // in case HEAD request failed for raw service, retry to get the image download metadata using thumbnail service
+      return await this.getImageDownloadMetadata();
```

Comment:
what if getImageDownloadMetadata rejects ?

Reply:
as we spoke, it returns a value